packages/contracts/contracts/universal/MultiSigWallet.sol

```diff
@@ -0,0 +1,529 @@
+    *
+    * @param _destination Transaction target address.
+    * @param _value Transaction ether value.
+    * @param _data Transaction data payload.
+    * @return Returns transaction ID.
+    */
+    function submitTransaction(address _destination, uint _value, bytes memory _data)
+        public
+        override
+        virtual
+        returns (uint)
+    {
+        uint transactionId = addTransaction(_destination, _value, _data);
+        confirmTransaction(transactionId);
+        return transactionId;
+    }
+
+    /**
+    * @notice Allows an owner to confirm a transaction.
+    *
+    * @param _transactionId Transaction ID.
+    */
+    function confirmTransaction(uint _transactionId)
+        public
+        override
+        virtual
+        ownerExists(msg.sender)
+        transactionExists(_transactionId)
+        notConfirmed(_transactionId, msg.sender)
+    {
+        confirmations[_transactionId][msg.sender] = true;
+        emit Confirmation(msg.sender, _transactionId);
+        _executeTransaction(_transactionId);
+    }
+
+    /**
+    * @notice Allows an owner to revoke a confirmation for a transaction.
+    *
+    * @param _transactionId Transaction ID.
+    */
+    function revokeConfirmation(uint _transactionId)
+        external
+        override
+        virtual
+        ownerExists(msg.sender)
+        confirmed(_transactionId, msg.sender)
+        notExecuted(_transactionId)
+    {
+        confirmations[_transactionId][msg.sender] = false;
+        emit Revocation(msg.sender, _transactionId);
+    }
+
+    /**
+    * @notice Internal functions. Execute a confirmed transaction.
+    *
+    * @param _transactionId Transaction ID.
+    */
+    function _executeTransaction(uint _transactionId)
+        internal
+        ownerExists(msg.sender)
+        confirmed(_transactionId, msg.sender)
+        notExecuted(_transactionId)
+    {
+        if (_isConfirmed(_transactionId)) {
+            Transaction storage txn = transactions[_transactionId];
+            txn.executed = true;
+            (bool success, ) = txn.destination.call(
+                txn.data
+            );
+            require(success, "call transaction failed");
+            emit Execution(_transactionId);
+        }
+    }
+
+    /**
+    * @notice Allows anyone to execute a confirmed transaction.
+    *
+    * @param _transactionId Transaction ID.
+    */
+    function executeTransaction(uint _transactionId)
+        external
+        override
+    {
+        _executeTransaction(_transactionId);
+    }
+
+    /**
+    * @notice Internal functions. Returns the confirmation status of a transaction.
+    *
+    * @param _transactionId Transaction ID.
+    * @return Confirmation status.
+    */
+    function _isConfirmed(uint _transactionId)
+        internal
+        view
+        returns (bool)
+    {
+        uint count = 0;
+        for (uint i = 0; i < owners.length; i++) {
+            if (confirmations[_transactionId][owners[i]])
```

Comment:
Please share the `confirmations[_transactionId]` part outside the loop.

Reply:
Applied.